### DISCUSSION TITLE
Editing items

### DISCUSSION THREAD

**Discussion by tdb** [zcutlip/pyonepassword#140] [Q&A]
How can I edit items? It looks like the `op` command itself supports this:

https://developer.1password.com/docs/cli/reference/management-commands/item/#item-edit

I suppose I could get the item, modify it, delete the original, then recreate it? But that doesn't seem like the right approach.

Maybe I'm missing something?

**Comment by zcutlip** [comment]
Hi, thanks for the question!

I've looked into adding this in the past, but it's not a great fit for `pyonepassword`. This project primarily works with `op` operations by consuming its JSON output such as `item get`, `user get`, etc. Or by feeding JSON to `op`'s stdin such as with `item create`, `item delete` (in the case of batch deletions), etc.

The way item editing is handled by `op` is by constructing a string as a positional CLI argument using a complex, nonstandard syntax to describe precisely what changes are desired. The edit string consists of a dotted namespace hierarchy to identify the field, as well as the datatype (one of several) for the new value.

Unfortunately this introduces a myriad of considerations that `pyonepassword` isn't currently designed to account for.

I would very much like to add at least some limited support for this, but I think it would be quite a reach at this stage.

If you are aware of any other projects similar to `pyonepassword` (in any language) that support item editing, please send me a note. I'd like to see how & if others are doing it for inspiration.

Zach

**Reply by zcutlip** [reply]
what would be nice is if `op` would let you take the JSON from an `op item get` operation, modify it then feed it to stdin for an `op item edit` operation

That would be easy for pyonepassword to support

**Reply by tdb** [reply]
I'm currently using https://github.com/1Password/connect-sdk-python, which is the 1Password Connect API. Obviously working completely differently under the hood, so likely not much help. I've been wanting to move to Service Accounts to avoid needing the 1Password Connect infrastructure though - my app is simple; it pulls an item, modifies the password, and updates it.

I was going to suggest feeding back to the developers, but I see you already posed the question on Slack :-)

**Reply by zcutlip** [reply]
If it's simply changing the password, that may be a common enough use case and straightforward enough I can expose API to to explicitly that without allowing doing a fully generic editing API

Would you mind sharing the JSON for an example item you wish to change the password on?

**Reply by tdb** [reply]
I over-simplified slightly, because I also explicitly save the old passwords in another field:

```
{
  "id": "itemid",
  "title": "item name",
  "version": 7,
  "vault": {
    "id": "vaultid",
    "name": "vault name"
  },
  "category": "LOGIN",
  "last_edited_by": "me",
  "created_at": "2022-07-29T20:54:40Z",
  "updated_at": "2023-07-20T15:39:58Z",
  "additional_information": "nothing",
  "fields": [
    {
      "id": "username",
      "type": "STRING",
      "purpose": "USERNAME",
      "label": "username",
      "value": "user",
      "reference": "op://vaultid/item name/username"
    },
    {
      "id": "password",
      "type": "CONCEALED",
      "purpose": "PASSWORD",
      "label": "password",
      "value": "password1",
      "entropy": 189.70021057128906,
      "reference": "op://vaultid/item name/password",
      "password_details": {
        "entropy": 189,
        "generated": true,
        "strength": "FANTASTIC",
        "history": [
          "password3",
          "password2"
        ]
      }
    },
    {
      "id": "idpw2",
      "type": "CONCEALED",
      "label": "password-20230720-163957",
      "value": "password2",
      "reference": "op://vaultid/item name/password-20230720-163957"
    },
    {
      "id": "idpw3",
      "type": "CONCEALED",
      "label": "password-20230720-162642",
      "value": "password3",
      "reference": "op://vaultid/item name/password-20230720-162642"
    },
  ]
}
```

So on a regular basis, I update the password in the password field, using the built-in generation. I then save the old password in a new field named with the date. Although passwords are saved in the password history, I also wanted them saved with the date.

A way to edit the username, password, and possibly title of an entry would likely satisfy a lot of use cases.

I'd like to better understand the complexity though. The documentation seems to suggest you can just do `op item edit 'My Example Item' 'field1[password]'`, which given the above would just be `op item edit 'item name' 'password[password]=newpass'`. Am I missing something?

It also looks like `item edit` doesn't use stdin, so comes with a warning about exposing passwords in the process list :-(

**Reply by zcutlip** [reply]
Okay, thanks for sharing, that helps.

I could see adding restricted edit operations like you describe.

The main problem lies in the fact that item objects and their fields can be arbitrarily complex, and aren't restricted to any particular schema. So translating between the CLI syntax and arbitrarily complex data types generically -- _and correctly_ -- is...well...not straightforward. If I do it wrong, and someone comes along with a weird object structure, I could end up corrupting or destroying user data which would be unacceptable.

But restricting edits to very specific field types and names I think might be manageable.

Adding additional fields for each saved password may not be in scope, but I'll look at that as well.

Yes, not taking edits over `stdin` is another problem which has held me off looking into this. There have been other features I didn't implement until `op` supported doing them securely. It's an important design goal of this project to not handle the user's data in an insecure way, _especially_ if that risk is not obvious. If I add this, I'll have to think of how to design the API in such a way that the user must acknowledge the risks.

Again, if you come across are any other `op` wrappers that are doing this even in other languages, please pass them on. I'd like to see how other projects are handling some of these issues.

**Comment by zcutlip** [comment]
Okay, got a tip-off in slack to look at @1password's official JS API, which supports item editing. So I'm going to study that and see what good ideas I can steal 😂
https://github.com/1Password/op-js

**Comment by zcutlip** [comment]
promoting this to an issue and see what I can work out for it.

**Comment by zcutlip** [comment]
I thought the entire discussion would be converted to an issue, but I guess not. Anyway, the issue can be tracked here: #143
